src/sage/geometry/cone_critical_angles.py

```diff
@@ -0,0 +1,998 @@
+        sage: K = random_cone(max_rays=8)
+        sage: G = matrix.column(K.rays())
+        sage: all( len(s) == K.rays(s).dimension() for s in gevp_licis(G) )
+        True
+    """
+    from sage.matroids.linear_matroid import LinearMatroid
+
+    # There's a fast implementation of this for matroids, but we need
+    # to drop the empty set from its output and convert the rest to
+    # lists that are all sorted in the same order.
+    return map(sorted, filter(bool,LinearMatroid(G).independent_sets()))
+
+
+def _solve_gevp_naive(GG, HH, M, I, J):
+    r"""
+    Solve the generalized eigenvalue problem in Theorem 3
+    [Or2020]_ in a very naive way, by (slowly) inverting the matrices
+    and finding the eigenvalues in the product space.
+
+    This is used only for testing, to ensure that the smart way of
+    solving the generalized eigenvalue problem via
+    :func:`solve_gevp_zero` and :func:`solve_gevp_nonzero` returns the
+    same answers as the dumb way.
+
+    This returns a generator, like :func:`solve_gevp_zero` and
+    :func:`solve_gevp_nonzero`.
+
+    INPUT:
+
+    See the arguments for :func:`solve_gevp_nonzero`.
+
+    ALGORITHM:
+
+    We construct the two matrices `A` and `B` in Theorem 3 in block
+    form, and then use the naive "inverse" method on `B` to move it to
+    the left and obtain `M = B^{-1}A`. We then compute the right
+    eigenvectors of the whole big matrix `M`.
+
+    EXAMPLES:
+
+    A simple usage example, that also appears as Example 3 in
+    [Or2020]_::
+
+        sage: from sage.geometry.cone_critical_angles import _solve_gevp_naive
+        sage: K = cones.nonnegative_orthant(2)
+        sage: G = matrix.column(K.rays())
+        sage: GG = G.transpose()*G
+        sage: I = [0]
+        sage: J = [1]
+        sage: list(_solve_gevp_naive(GG,GG,GG,I,J))
+        [(0, (1), (0), 2), (0, (0), (1), 2)]
+
+    Check Example 4 [Or2020]_ symbolically to ensure that we get
+    eigenspaces of dimension `n=2` corresponding to the eigenvalues
+    `\cos\theta = -1` and `\cos\theta = 1`::
+
+        sage: from sage.geometry.cone_critical_angles import _solve_gevp_naive
+        sage: g11,g12,g21,g22 = SR.var('g11,g12,g21,g22', domain='real')
+        sage: h11,h12,h21,h22 = SR.var('h11,h12,h21,h22', domain='real')
+        sage: gs = [[g11,g12], [g21,g22]]
+        sage: hs = [[h11,h12], [h21,h22]]
+        sage: G = matrix.column(gs)
+        sage: H = matrix.column(hs)
+        sage: GG = G.transpose()*G
+        sage: HH = H.transpose()*H
+        sage: M = G.transpose()*H
+        sage: I = [0, 1]
+        sage: J = [0, 1]
+        sage: all( v in [-1,1] and m == 2
+        ....:      for (v,_,_,m) in _solve_gevp_naive(GG,HH,M,I,J) )
+        True
+    """
+    A = matrix.block([ [ ZZ(0),              M[I,J] ],
+                       [ M.transpose()[J,I], ZZ(0)  ] ])
+    B = matrix.block([ [ GG[I,I], ZZ(0)   ],
+                       [ ZZ(0),   HH[J,J] ] ])
+    M = B.inverse() * A
```

Comment:
```suggestion
    A = matrix.block(ZZ, [[ ZZ.zero(), M[I,J] ],
                       [M[I,J].transpose(), ZZ.zero()]])
    Binv = matrix.block_diagonal(ZZ, [GG[I,I].inverse(), HH[J,J].inverse()])
    M = Binv * A
```
These are mostly mirco-optimizations with more direct cache usage. It is better to transpose and invert the smaller matrices too. PEP8 also doesn't like such alignments... (yes, I think there should be exceptions too...)

Reply:
I've changed `ZZ(0)` to `ZZ.zero()` since that is a good idea in any case. But this code is inside the `_solve_gevp_naive` function, whose only purpose is to test the two non-naive GEVP functions. As the docs say,

> This is used only for testing, to ensure that the smart way of solving the generalized eigenvalue problem ... returns the same answers as the dumb way.

For that reason I have avoided using any mathematical tricks (no matter how obvious) to simplify the problem. Is it OK to leave the block matrices? I can trim some of the spaces to conform to PEP8.

Reply:
It's not really a trick, it's simply realizing it is a block diagonal matrix. Anyways, if you prefer it, it is okay.

Reply:
Ok, thanks. I've removed the extra spacing and reindented:

```python
    A = matrix.block([
        [ZZ.zero(), M[I,J]],
        [M.transpose()[J,I], ZZ.zero()]
    ])
    B = matrix.block([
        [GG[I,I], ZZ.zero()],
        [ZZ.zero(), HH[J,J]]
    ])
```

PEP8 is okay with that.